x-pack/plugins/enterprise_search/public/applications/shared/flash_messages/types.ts

```diff
@@ -5,10 +5,18 @@
  * 2.0.
  */
 
-import { ReactNode } from 'react';
+import { ReactNode, ReactChild } from 'react';
 
 export interface IFlashMessage {
   type: 'success' | 'info' | 'warning' | 'error';
   message: ReactNode;
   description?: ReactNode;
 }
+
+// @see EuiGlobalToastListToast
+export interface IToast {
+  id: string;
+  title?: ReactNode;
+  text?: ReactChild;
+  toastLifeTimeMs?: number;
+}
```

Comment:
This is basically just a copy of EUI's `Toast` obj/data structure. Unfortunately while it's available for inspection in `eui.d.ts` & they list in their docs, they don't actually expand on what's in it:

<img width="555" alt="Screen Shot 2021-03-31 at 1 31 44 PM" src="https://user-images.githubusercontent.com/549407/113207503-8a7fab00-9225-11eb-9f0b-6298b71aa490.png">

So I thought I would essentially re-iterate it here to improve our developer experience/documentation.

Reply:
Also FYI, I believe `id` is required for accessibility reasons. It confuses me a bit that `title` is _not_ required, but I get Typescript errors if I try to make it required, so I left it as-is

Reply:
> Is there more to it than "the docs don't explicitly explain what a Toast is?"

There's slightly more to it, sorry. Here's the errors we get when specifically referencing `EuiGlobalToastListToast`:

```
 x-pack/plugins/enterprise_search/public/applications/shared/flash_messages/set_message_helpers.ts:45:46 - error TS2345: Argument of type '{ id: string; title?: React.ReactNode; text?: string | number | React.ReactElement<any, string | ((props: any) => React.ReactElement<any, string | ... | (new (props: any) => React.Component<any, any, any>)> | null) | (new (props: any) => React.Component<...>)> | undefined; toastLifeTimeMs?: number | undefined; color...' is not assignable to parameter of type 'Toast'.
  Types of property 'color' are incompatible.
    Type 'string' is not assignable to type '"success" | "danger" | "warning" | "primary" | undefined'.

45   FlashMessagesLogic.actions.addToastMessage({
                                                ~
46     ...FLASH_MESSAGE_TYPES.success,
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

It's super annoying and it's basically a similar issue in the file we have to cast `EuiCallOutProps['color']` because it doesn't recognize that we're passing in an allowed string due to it being an obj property. I wish there were like a single `EuiAllowedColors` enum we could use or something, but I don't think anything like that exists from EUI 🤔

Reply:
Oh, I should also expand slightly that `EuiGlobalToastListToast` allows `color` and `iconType` props that we don't document in `IToast`, because we set that automatically and as such we don't expect _our_ usage of our own toast API to have those props. So I guess maybe that also helps validate `IToast` as being a separate (but similar) entity?

Looking at `EuiGlobalToastListToast` and `EuiToastProps` more closely, it looks like it also expects an `onClose` fn that we don't require since it's passed automatically by EUI.

Reply:
I was able to fix the color type errors in https://github.com/elastic/kibana/pull/95981/commits/78255f5e3a7558330fe0d6f094a31d7e7245c2d5 and also reference `EuiGlobalToastListToast` directly in FlashMessagesLogic.

I kept our own custom toast type definition in as our `id` is optional in our helper compared to EUI's Toast, and I really do think having the types/options listed here is helpful to other developers to make up for a gap in EUI's documentation, but definitely open to discussing more.